optimize/local.go

```diff
@@ -4,222 +4,139 @@
-//  method := &optimize.BFGS{}
-// creates a pointer to a new BFGS struct. When Local is called, the settings
-// in the method will be populated with default values. The methods are also
-// designed such that they can be reused in future calls to Local.
-//
-// If method implements Statuser, method.Status is called before every call
-// to method.Iterate. If the returned Status is not NotTerminated or the
-// error is non-nil, the optimization run is terminated.
-//
-// Local returns a Result struct and any error that occurred. See the
-// documentation of Result for more information.
-//
-// Be aware that the default behavior of Local is to find the minimum.
-// For certain functions and optimization methods, this process can take many
-// function evaluations. If you would like to put limits on this, for example
-// maximum runtime or maximum function evaluations, modify the Settings
-// input struct.
-func Local(p Problem, initX []float64, settings *Settings, method Method) (*Result, error) {
-	if method == nil {
-		method = getDefaultMethod(&p)
-	}
-	if settings == nil {
-		settings = DefaultSettings()
-	}
-	// Check that the initial location matches the one in settings.
-	if settings.InitX != nil && !floats.Equal(settings.InitX, initX) {
-		panic("local: initX does not match settings x location")
-	}
-	lg := &localGlobal{
-		Method:   method,
-		InitX:    initX,
-		Settings: settings,
-	}
-	return Global(p, len(initX), settings, lg)
-}
-
-func getDefaultMethod(p *Problem) Method {
-	if p.Grad != nil {
-		return &BFGS{}
-	}
-	return &NelderMead{}
-}
-
-// localGlobal is a wrapper for Local methods to allow them to be optimized by Global.
-type localGlobal struct {
-	Method   Method
-	InitX    []float64
-	Settings *Settings
-
-	dim    int
 	status Status
 	err    error
 }
 
-func (l *localGlobal) InitGlobal(dim, tasks int) int {
-	if dim != len(l.InitX) {
-		panic("optimize: initial length mismatch")
-	}
-	l.dim = dim
-	l.status = NotTerminated
-	l.err = nil
-	return 1 // Local optimizations always run in serial.
-}
-
-func (l *localGlobal) Status() (Status, error) {
+func (l *LocalController) Status() (Status, error) {
 	return l.status, l.err
 }
 
-func (l *localGlobal) Needs() struct {
-	Gradient bool
-	Hessian  bool
-} {
-	return l.Method.Needs()
+func (l *LocalController) InitGlobal(dim, tasks int) int {
+	l.status = NotTerminated
+	l.err = nil
+	return 1 // Local optimizations always run in serial.
 }
 
-func (l *localGlobal) RunGlobal(operations chan<- GlobalTask, results <-chan GlobalTask, tasks []GlobalTask) {
+func (l *LocalController) RunGlobal(operation chan<- GlobalTask, result <-chan GlobalTask, tasks []GlobalTask) {
 	// Local methods start with a fully-specified initial location.
 	task := tasks[0]
-	op := l.getStartingLocation(operations, results, task)
-	if op == PostIteration {
-		l.cleanup(operations, results)
+	task = l.initialLocation(operation, result, task, l.Method)
```

Comment:
I changed to `localOptimizer` and made the type completely stateless. This way it can provide helper routines when/if we transform the local methods, and having the type helps keep the name space clean.